File: data-pipeline/scripts/backfill_bills.py
"""Incrementally backfill historical Congress.gov passage-action bills.

Run on a daily schedule (separate from the fresh-pull workflow). Each
invocation processes a bounded number of list pages for the active Congress
in the backfill queue, merges the results into that Congress's per-Congress
manifest, advances state, and rewrites the index. Failures leave the cursor
where it was so the next run retries the same offset.
"""

from __future__ import annotations

import os
import sys
from collections import Counter
from datetime import datetime, timezone
from typing import Any

from _common import (
    BACKFILL_PAGES_PER_RUN,
    CongressClient,
    LIST_PAGE_LIMIT,
    advance_state,
    build_bill_record,
    evaluate_bill,
    load_manifest,
    load_state,
    merge_records,
    rebuild_index,
    save_manifest,
    save_state,
)

# Effectively no date floor — backfill keeps every passage-action bill we
# find, regardless of how long ago the action was.
NO_DATE_CUTOFF = datetime(1970, 1, 1, tzinfo=timezone.utc)


def list_congress_page(
    client: CongressClient, congress: int, offset: int
) -> list[dict[str, Any]]:
    body = client.get(
        f"/bill/{congress}",
        limit=LIST_PAGE_LIMIT,
        offset=offset,
        sort="updateDate asc",
    )
    return body.get("bills") or []


def main() -> int:
    api_key = os.environ.get("CONGRESS_API_KEY")
    if not api_key:
        print("CONGRESS_API_KEY is not set in the environment.", file=sys.stderr)
        return 2

    state = load_state()
    active = state.get("active_congress")
    if active is None:
        print("Backfill complete: queue is empty.")
        rebuild_index()
        return 0

    offset = state.get("active_offset", 0)
    print(
        f"Backfilling Congress {active} starting at offset {offset} "
        f"(pages_per_run={BACKFILL_PAGES_PER_RUN})"
    )

    client = CongressClient(api_key)
    fresh_records: list[dict[str, Any]] = []
    seen_ids: set[str] = set()
    reject_counts: Counter[str] = Counter()
    total_evaluated = 0
    last_page_size = LIST_PAGE_LIMIT
    pages_consumed = 0
    had_non_empty_page = False

    for page in range(BACKFILL_PAGES_PER_RUN):
        page_offset = offset + page * LIST_PAGE_LIMIT
        bills = list_congress_page(client, active, page_offset)
        last_page_size = len(bills)
        pages_consumed += 1
        if last_page_size > 0:
            had_non_empty_page = True

        for summary in bills:
            total_evaluated += 1
            outcome, reason = evaluate_bill(summary, NO_DATE_CUTOFF)
            if outcome is None:
                reject_counts[reason] += 1
                continue
            try:
                record = build_bill_record(client, active, summary, outcome)
            except Exception as exc:  # noqa: BLE001 - one bad bill must not kill the run
                ref = f"{summary.get('type')}{summary.get('number')}"
                print(
                    f"  ! skipping {ref}: {type(exc).__name__}: {exc}",
                    file=sys.stderr,
                )
                reject_counts["build_error"] += 1
                continue
            if record["id"] in seen_ids:
                reject_counts["duplicate"] += 1
                continue
            seen_ids.add(record["id"])
            fresh_records.append(record)

        if last_page_size < LIST_PAGE_LIMIT:
            break

    print(
        f"Evaluated {total_evaluated} bills across {pages_consumed} page(s); "
        f"kept {len(fresh_records)}."
    )
    if reject_counts:
        print("Rejections:")
        for reason, count in reject_counts.most_common():
            print(f"  - {reason}: {count}")

    existing = load_manifest(active)
    merged, stats = merge_records(existing.get("bills", []), fresh_records)
    final = save_manifest(active, {"bills": merged})

    new_state = advance_state(
        state, last_page_size, pages_consumed, had_non_empty_page
    )
    save_state(new_state)
    rebuild_index()

    print(
        f"merge: +{stats.added} added, ~{stats.updated} updated, "
        f"={stats.unchanged} unchanged (manifest now {len(final['bills'])} bills)"
    )
    held_cursor = (
        new_state.get("active_congress") == active
        and new_state.get("active_offset") == offset
    )
    if last_page_size < LIST_PAGE_LIMIT and not held_cursor:
        print(
            f"Congress {active} complete; "
            f"next: {new_state.get('active_congress')}"
        )
    elif held_cursor and last_page_size < LIST_PAGE_LIMIT:
        print(
            f"Empty page at offset {offset} with no prior progress; "
            f"treating as transient and holding cursor on Congress {active}."
        )
    else:
        print(
            f"Next run resumes at Congress {active} "
            f"offset {new_state['active_offset']}"
        )
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: pipeline/shared/src/commonMain/kotlin/com/informedcitizen/pipeline/fetch/BackfillBills.kt
package com.informedcitizen.pipeline.fetch

import com.informedcitizen.pipeline.ErrorCollector
import com.informedcitizen.pipeline.http.CongressClient
import com.informedcitizen.pipeline.model.Bill
import com.informedcitizen.pipeline.model.BillsManifest
import com.informedcitizen.pipeline.state.BACKFILL_PAGES_PER_RUN
import com.informedcitizen.pipeline.state.BackfillState
import com.informedcitizen.pipeline.state.LIST_PAGE_LIMIT
import com.informedcitizen.pipeline.state.advanceBackfillState
import kotlinx.datetime.Instant
import kotlinx.serialization.json.JsonObject

/**
 * Effectively no date floor — the backfill keeps every passage-action
 * bill it finds, regardless of how long ago the action was. Python:
 * `NO_DATE_CUTOFF = datetime(1970, 1, 1, tzinfo=timezone.utc)`.
 */
private val BACKFILL_NO_DATE_CUTOFF: Instant = Instant.fromEpochSeconds(0)

/** Summary of a single backfill run, returned by [backfillBills]. */
data class BackfillBillsResult(
    val activeCongress: Int?,
    val priorOffset: Int,
    val evaluated: Int,
    val keptRecords: List<Bill>,
    val pagesConsumed: Int,
    val lastPageSize: Int,
    val hadNonEmptyPage: Boolean,
    val rejectionCounts: Map<String, Int>,
    val mergeStats: MergeStats?,
    val finalManifest: BillsManifest?,
    val newState: BackfillState,
    /** True iff the queue was already exhausted at run start; nothing was fetched. */
    val queueWasEmpty: Boolean,
    /** True iff the empty-page guard fired and the cursor stayed put. */
    val cursorHeld: Boolean,
    /** True iff this run finished the active Congress (short page + had evidence). */
    val congressCompleted: Boolean,
)

/**
 * Historical backfill orchestrator. Direct port of Python
 * `backfill_bills.main`'s flow:
 *
 *  1. If `state.activeCongress == null`, the queue is exhausted: return
 *     early with [BackfillBillsResult.queueWasEmpty] = true. Caller
 *     should NOT save state in this case (matches Python's "don't
 *     bump `last_run_at` on a no-op run").
 *  2. Otherwise, walk up to [pagesPerRun] pages of `/bill/{congress}`
 *     starting at `state.activeOffset`, sorted ascending by
 *     `updateDate`. Apply [evaluateBill] with no date cutoff —
 *     historical bills of any age are kept if they match an outcome
 *     rule.
 *  3. Enrich kept bills in parallel via [buildBillRecordsParallel].
 *  4. Dedupe by id.
 *  5. Merge into the existing manifest for the active Congress.
 *  6. Persist the merged manifest.
 *  7. Compute the new state via [advanceBackfillState] (which
 *     preserves the empty-page guard from the 2026-05-05 fix —
 *     transient empty first pages at offset 0 leave the cursor in
 *     place instead of marking the Congress complete).
 *  8. Return the new state plus run summary; caller persists.
 *
 * `congresses.json` rebuild is NOT performed here — same scope
 * boundary as [fetchBills]. The CI workflow keeps the Python
 * `build_dashboard.py` step until that builder is ported.
 */
suspend fun backfillBills(
    client: CongressClient,
    state: BackfillState,
    nowIso: String,
    manifestStore: FileBillsManifestStore,
    errors: ErrorCollector,
    pagesPerRun: Int = BACKFILL_PAGES_PER_RUN,
    maxWorkers: Int = ENRICH_WORKERS,
): BackfillBillsResult {
    val active = state.activeCongress
    if (active == null) {
        return BackfillBillsResult(
            activeCongress = null,
            priorOffset = state.activeOffset,
            evaluated = 0,
            keptRecords = emptyList(),
            pagesConsumed = 0,
            lastPageSize = 0,
            hadNonEmptyPage = false,
            rejectionCounts = emptyMap(),
            mergeStats = null,
            finalManifest = null,
            newState = state,
            queueWasEmpty = true,
            cursorHeld = false,
            congressCompleted = false,
        )
    }

    val priorOffset = state.activeOffset
    val rejectCounts = mutableMapOf<String, Int>()
    val keptSummaries = mutableListOf<Pair<JsonObject, String>>()
    var totalEvaluated = 0
    var lastPageSize = LIST_PAGE_LIMIT
    var pagesConsumed = 0
    var hadNonEmptyPage = false

    // Phase 1: paginate sequentially. Filter on the way through.
    for (page in 0 until pagesPerRun) {
        val pageOffset = priorOffset + page * LIST_PAGE_LIMIT
        val bills = listCongressPage(client, active, pageOffset)
        lastPageSize = bills.size
        pagesConsumed++
        if (lastPageSize > 0) hadNonEmptyPage = true

        for (summary in bills) {
            totalEvaluated++
            when (val result = evaluateBill(summary, BACKFILL_NO_DATE_CUTOFF)) {
                is BillEvaluationResult.Kept -> keptSummaries += summary to result.outcome
                is BillEvaluationResult.Rejected -> rejectCounts.bump(result.reason)
            }
        }

        if (lastPageSize < LIST_PAGE_LIMIT) break
    }

    // Phase 2: parallel enrichment.
    val (freshRecords, buildFailures) =
        buildBillRecordsParallel(client, active, keptSummaries, errors, maxWorkers)
    if (buildFailures > 0) rejectCounts.bump(RejectionReasons.BUILD_ERROR, buildFailures)

    // Dedupe by id (rare cross-page repeats).
    val seenIds = mutableSetOf<String>()
    val deduped = mutableListOf<Bill>()
    for (rec in freshRecords) {
        if (rec.id in seenIds) {
            rejectCounts.bump(RejectionReasons.DUPLICATE)
            continue
        }
        seenIds += rec.id
        deduped += rec
    }

    // Merge into the existing manifest and persist.
    val existing = manifestStore.load(active)?.bills.orEmpty()
    val (merged, mergeStats) = mergeBillRecords(existing, deduped)
    val finalManifest = manifestStore.save(active, merged, nowIso)

    // Advance state. The empty-page guard inside `advanceBackfillState`
    // distinguishes "real exhaustion" (we saw evidence) from "transient
    // empty page at offset 0 with nothing to back it up" (hold cursor).
    val newState = advanceBackfillState(
        state = state,
        pageReturned = lastPageSize,
        pagesConsumed = pagesConsumed,
        hadNonEmptyPage = hadNonEmptyPage,
        nowIso = nowIso,
    )

    // `cursorHeld` ⇒ the empty-page guard fired; we stayed on the same
    // (congress, offset). `congressCompleted` ⇒ `advanceBackfillState`
    // moved `active` into `completed`. Mutually exclusive; the third
    // case (offset bumped within the same congress) is the default.
    val cursorHeld =
        newState.activeCongress == active && newState.activeOffset == priorOffset
    val congressCompleted =
        active in newState.completed && active !in state.completed

    return BackfillBillsResult(
        activeCongress = active,
        priorOffset = priorOffset,
        evaluated = totalEvaluated,
        keptRecords = deduped,
        pagesConsumed = pagesConsumed,
        lastPageSize = lastPageSize,
        hadNonEmptyPage = hadNonEmptyPage,
        rejectionCounts = rejectCounts.toMap(),
        mergeStats = mergeStats,
        finalManifest = finalManifest,
        newState = newState,
        queueWasEmpty = false,
        cursorHeld = cursorHeld,
        congressCompleted = congressCompleted,
    )
}

// `MutableMap.merge` is JVM-only — duplicate of the helper in FetchBills.kt
// since kept private there. Trivial enough to repeat rather than promote.
private fun MutableMap<String, Int>.bump(key: String, by: Int = 1) {
    this[key] = (this[key] ?: 0) + by
}
